Use SingleR to annotate SCPCS000492

allyhawkins opened the pull request:
### Purpose/implementation Section

#### Please link to the GitHub issue that this pull request addresses.

Closes #544 

#### What is the goal of this pull request?

Here I explored using `SingleR` to classify tumor cells in `SCPCS000492`. I am specifically focused on identifying tumor cells rather than individual normal cell types (endothelial cells, etc). 

#### Briefly describe the general approach you took to achieve this goal.

Running `SingleR`: 

- I wrote a script that runs `SingleR` using three different reference types. The input is a reference SCE file, in this case `SCPCL000822` and the output is classifications obtained by using three different modifications of the input reference: 
  - `Blueprint`: Here we used `SCPCL000822` with cells labeled as either tumor or normal in combination with `BlueprintEncodeData`, which was the original reference we used with `scpca-nf`. 
  - `Original`: Here we used `SCPCL000822` with the original `SingleR` annotations, but replaced any tumor cells we identified with the tumor class label. 
  - `Original + Blueprint`: Here we used the same thing as the above bullet where tumor cells replaced original labels and combined with `BlueprintEncodeData`. 
- The output for all three methods are saved to a TSV file in the results folder and read into the exploratory notebook. 

Exploring the results: 

- I created a notebook that reads in the `SingleR` results and the marker gene annotations obtained in #532. 
- The cell type annotations for each of the references are displayed in UMAPs similar to what we show in the cell type report. 
- We compare the cells that are identified as tumor across all three methods. Generally they are the same, with some cells identified only by one or two of the references. 
- The normal cell types are compared to the original cell types identified by `SingleR` in `scpca-nf`. 
- The cells that are classified as tumor are compared across all `SingleR` refs and marker gene based annotations. 
- And finally we look at marker gene expression and gene set scores in tumor cells identified by `SingleR`. 

#### If known, do you anticipate filing additional pull requests to complete this analysis module?

Yes 


### Results

#### What is the name of your results bucket on S3?

`s3://researcher-211125375652-us-east-2/cell-type-ewings/results/annotation_tables/SCPCS000492`

#### What types of results does your code produce (e.g., table, figure)?

The `SingleR` script outputs both the `SingleR` objects (which are currently not really used, but I saved them just in case) and the TSV file with all results from `SingleR`, `SCPCL000824_singler-classifications.tsv.gz`. 

#### What is your summary of the results?

 - For the most part, there is agreement in the cells that are called as tumor cells regardless of the reference used. However, using `SCPCL000822` with tumor and normal cells as labels and Blueprint seemed to show the highest consistency with marker gene based methods and biggest separation of marker gene expression and gene set scores between tumor and all other cells. This makes me think for annotating tumor cells, we should use this reference. 
 - There is very little agreement between the normal cell types identified with these references and the original normal annotations from `scpca-nf`. This is maybe not super surprising since tumor cells were probably mis labeled, but I did remove those before doing the comparison. I think we will need to think about how to confidently annotate normal cell types once we can confidently identify tumor cells. 

Here is the rendered notebook: 
[SCPCL000824_singler-classification.html.zip](https://github.com/user-attachments/files/16071937/SCPCL000824_singler-classification.html.zip)


### Provide directions for reviewers

#### What are the software and computational requirements needed to be able to run the code in this PR?

I ran everything locally with 4 threads. 


#### Are there particularly areas you'd like reviewers to have a close look at?

One thing I thought of later on is to pull out just the tumor cells from `SCPCL000822` and combine with `Blueprint`. Do we think this is something worth doing? 

Also, I wasn't totally sure where to put the script for `SingleR`. Right now it's in the `scripts` directory, but it's pretty specific to this analysis, so I wonder if we should move it. 


#### Is there anything that you want to discuss further?

I am filing this as a draft for now since I'm waiting for #532 to go in first. 
Once that's in I'll update this and request formal review. 



### Author checklists

_Check all those that apply._
_Note that you may find it easier to check off these items after the pull request is actually filed._


#### Analysis module and review

- [ ] This analysis module [uses the analysis template and has the expected directory structure](https://openscpca.readthedocs.io/en/latest/contributing-to-analyses/analysis-modules/).
- [x] The analysis module `README.md` has been updated to reflect code changes in this pull request.
- [x] The analytical code is documented and contains comments.
- [x] Any results and/or plots this code produces have been added to your S3 bucket for review.

#### Reproducibility checklist

- [ ] Code in this pull request has been added to the GitHub Action workflow that runs this module.
- [ ] The dependencies required to run the code in this pull request have been added to the analysis module `Dockerfile`.
- [ ] If applicable, the dependencies required to run the code in this pull request have been added to the analysis module conda `environment.yml` file.
- [x] If applicable, R package dependencies required to run the code in this pull request have been added to the analysis module `renv.lock` file.

jashapiro commented:
> * `Blueprint`: Here we used `SCPCL000822` with cells labeled as either tumor or normal in combination with `BlueprintEncodeData`, which was the original reference we used with `scpca-nf`.
> * `Original`: Here we used `SCPCL000822` with the original `SingleR` annotations, but replaced any tumor cells we identified with the tumor class label.
> * `Original + Blueprint`: Here we used the same thing as the above bullet where tumor cells replaced original labels and combined with `BlueprintEncodeData`.

Can you put these explanations in the notebook? I was confused for a bit looking just at the notebook about why there were cell type calls and "Normal" calls in some of these.

jashapiro commented:
> Can you put these explanations in the notebook? I was confused for a bit looking just at the notebook about why there were cell type calls and "Normal" calls in some of these.

Ugh, never mind, it is there. I just didn't see it with the Setup between it and the plots.

allyhawkins commented:
Noting that this is now ready for review.

jashapiro reviewed (commented):
Overall I think this looks good. My main "operational" critique is that I want to see the commands that you used to invoke the script you wrote to generate the output. You will want that too if you have to rerun things! (I think the script location is just fine). There are a few little tweaks too, but nothing major there. 

Content-wise, I think this is in a pretty good place. I have a number of musing below, but I am not really sure which of those would really be most productive to pursue, if any. 

It looks to me like the best option is the `Original and Blueprint`; the Blueprint only option seems to be "absorbing" the bad calls into the broader "Normal" category, which seems dangerous to me as that is a clearly heterogeneous population that seems like it could have odd effects. Which may be what you are seeing when it has the most Tumor calls that are unique to that method.

One thought I had there was that maybe we could cluster the Normal cells and use those as a reference with more homogeneity within groups. I have not fully formed this thought, and I am not sure it would do any better than the `Original and Blueprint` method. 

I also had some thoughts that we might want to do the `Original and Blueprint` while giving the 822 samples distinct cell-type labels (i.e. `epithelial-822`). Not sure what the effect of that would be, but I guess I am curious whether 824 cells would better match the 822 or blueprint patterns, and this would allow you to see that (though maybe you can see it in the results as calculated directly too? Not sure if the reference matched is retained in the output at all for multiple refs)

One more thought I had was to see what happens if you use the cellassign + tumor classifications and then try applying those with singleR. This also seems not great though.

I _don't_ think you will want to make a reference that is only tumor cells (+ Blueprint), as I would tend to expect that to be more biased, but I don't think I have a good enough intuition to rule it out either. So maybe try it?

Just to be clear (again) the reference tumor assignments here are determined by copy number? Do you get something different if you use the marker gene analysis/AUCell calls as the original classifier?

allyhawkins commented:
@jashapiro thanks for this feedback. I addressed most of the inline comments including: 

- Adding a shell script to run the SingleR script specifically for this sample. I'm saving it in the same folder as the notebook. 
- For the NA values in the plot I ended up making an adjustment so that they are considered a group and get grouped with "All remaining cell types". They still show up in the density plots looking at gene expression, but I think that's okay. There aren't very many of them so I don't think it was worth showing on the UMAP. 

Here's an updated report: 
[SCPCL000824_singler-classification.html.zip](https://github.com/user-attachments/files/16112624/SCPCL000824_singler-classification.html.zip)

> It looks to me like the best option is the `Original and Blueprint`; the Blueprint only option seems to be "absorbing" the bad calls into the broader "Normal" category, which seems dangerous to me as that is a clearly heterogeneous population that seems like it could have odd effects. Which may be what you are seeing when it has the most Tumor calls that are unique to that method.

I mostly agree with this when looking at the SingleR methods alone and have updated some concluding text in that section to reflect that. However, when looking at the marker gene expression and gene set scores, Blueprint only tends to have the biggest separation between tumor and all other cells. So I do think with the other references we may be missing some tumor calls. Blueprint alone actually looks more similar to the marker gene based approaches. 

> One thought I had there was that maybe we could cluster the Normal cells and use those as a reference with more homogeneity within groups. I have not fully formed this thought, and I am not sure it would do any better than the `Original and Blueprint` method.

I think I see where you are going with this, but I don't know if it's worth the effort at this point. I agree that it probably won't do much better. 

> I also had some thoughts that we might want to do the `Original and Blueprint` while giving the 822 samples distinct cell-type labels (i.e. `epithelial-822`). Not sure what the effect of that would be, but I guess I am curious whether 824 cells would better match the 822 or blueprint patterns, and this would allow you to see that (though maybe you can see it in the results as calculated directly too? Not sure if the reference matched is retained in the output at all for multiple refs)

I added a small section that looks at the original labels for the "Original and Blueprint". I think some of the top cell types do identify more with cells in 822 over the celldex reference (endothelial and chondrocytes), while others are more similar to celldex (fibroblasts and neurons). Because of this I think I like keeping in both the original labels and the celldex reference. Again, I don't expect the groups of normal cells to be entirely equivalent across samples so I think we need to keep providing celldex. 

> I _don't_ think you will want to make a reference that is only tumor cells (+ Blueprint), as I would tend to expect that to be more biased, but I don't think I have a good enough intuition to rule it out either. So maybe try it?

Since this was easy to implement I went ahead and tried this. I think you're right about it being more biased and generally most cells are called tumor here. I think we probably don't want to do this moving forward. 

> Just to be clear (again) the reference tumor assignments here are determined by copy number? Do you get something different if you use the marker gene analysis/AUCell calls as the original classifier?

Yes, these were annotations we obtained from CNV inference that also matched up with marker gene classification. The confusion matrices that we calculated comparing the CNV assignment to marker genes/ AUCell showed almost complete agreement. So I don't think this would change the outcome that much. 

In thinking about this more, I am a little concerned about using `SingleR` with this one sample as a reference for tumor cells to classify tumor cells in all other samples. With only one reference set of tumor cells whatever specific genes that are marker genes for those tumor cells in that sample are what will be used to identify tumor cells in other samples. Ewing's samples are very heterogeneous, so using just one sample does concern me. With using a marker gene based approach we have curated a list of genes we expect to be found in all tumor cells, regardless of sample, and then are using that to define the set of tumor cells. I think I would be more enthusiastic about using `SingleR` if we had multiple samples with defined tumor cells as the reference. 

Because of that, I think we may want to actually use `AUCell` or other marker gene based approach to classify tumor cells in all of our samples. Then we can take a few samples that we feel very confident in and use that as a reference with Blueprint and run `SingleR` to see if we can refine the tumor cell calls. I haven't fully formed this thought yet and need to think about it some more, but I am leaning towards starting with just marker gene based approaches.

jashapiro commented:
> In thinking about this more, I am a little concerned about using `SingleR` with this one sample as a reference for tumor cells to classify tumor cells in all other samples. With only one reference set of tumor cells whatever specific genes that are marker genes for those tumor cells in that sample are what will be used to identify tumor cells in other samples. Ewing's samples are very heterogeneous, so using just one sample does concern me. With using a marker gene based approach we have curated a list of genes we expect to be found in all tumor cells, regardless of sample, and then are using that to define the set of tumor cells. I think I would be more enthusiastic about using `SingleR` if we had multiple samples with defined tumor cells as the reference.
> 
> Because of that, I think we may want to actually use `AUCell` or other marker gene based approach to classify tumor cells in all of our samples. Then we can take a few samples that we feel very confident in and use that as a reference with Blueprint and run `SingleR` to see if we can refine the tumor cell calls. I haven't fully formed this thought yet and need to think about it some more, but I am leaning towards starting with just marker gene based approaches.

Before I got through a full review, I do want to say that I endorse everything you said here. I think it is interesting to explore what happens with SingleR using a single reference sample as you do in this notebook, but my inutiution too is that I would not want to rely on it.

jashapiro reviewed (approved):
I had a couple of minor comments below, but I think my main comment remains the one I left prior to this, which is to say that I agree with your assessment that you should start with the marker-gene based annotation methods and maybe return to SingleR when we have a more diverse set of tumor samples to work from. 

So I think this notebook is in a good place for an exploration, and it makes sense to move on to the next steps!

allyhawkins commented on analyses/cell-type-ewings/exploratory_analysis/annotation_notebooks/SCPCS000492/SCPCL000824_singler-classification.Rmd:
```diff
@@ -0,0 +1,539 @@
+These same cells match to their corresponding Blueprint cell type when combining `SCPCL000822` and Blueprint (e.g., fibroblasts, neurons, and epithelial cells are identifed in both). 
+- When comparing using just `SCPCL000822` (original) to `SCPCL000822` and Blueprint (original and blueprint), we see an overlap in endothelial cells and chrondrocytes. 
+We also see that for the most part the top cell types identified are the same except when using blueprint you see the addition of epithelial cells. 
+- Using tumor only appears to label some cells as tumor that are labeled as normal with other methods. 
+For example, comparing tumor only to both original and original and blueprint shows that tumor cells are classified as endothelial cells using either of the other two methods.  
```
Do you mean whether or not the tumor cells are correct? Or if the calls for them being endothelial cells are correct? I think multiple methods calling them endothelial cells (SingleR, CellAssign, and then using our custom refs here) makes me trust them to be endothelial over tumor?

jashapiro replied:
I do mean whether calling them tumor or not is correct.

My question here is really whether the "tumor only" ref is catching extra true tumor cells or if those cells were being correctly assigned by the "original" method.  From the UMAPs it did look like the "tumor only" was calling more cells outside the lower group which I would have expected were false positives, but I wasn't sure if that was a reasonable conclusion. I don't think it is something we have to know at this point, but I wasn't sure if you had a sense.

allyhawkins replied:
I don't know for sure, but one way to look at that is specifically pull out those cells and see if they are called tumor by marker gene methods. I think that's the only way I would really trust that they are tumor at this point. My sense is that they are not since on the marker gene UMAPs they are also considered normal cells.
